osu.Game/Overlays/SocialOverlay.cs

```diff
@@ -62,57 +57,146 @@ public SocialOverlay()
+                panels.FadeOut(200);
+                panels.Expire();
+                panels = null;
+            }
 
-            // no this is not the correct data source, but it's something.
-            var request = new GetUsersRequest();
-            request.Success += res =>
+            if (Users == null)
+                return;
+
+            var newPanels = new FillFlowContainer<UserPanel>
             {
-                Users = res.Select(e => e.User);
-                loading.Hide();
+                RelativeSizeAxes = Axes.X,
+                AutoSizeAxes = Axes.Y,
+                Spacing = new Vector2(10f),
+                Margin = new MarginPadding { Top = 10 },
+                ChildrenEnumerable = Users.Select(u =>
+                {
+                    UserPanel panel = new UserPanel(u)
+                    {
+                        Anchor = Anchor.TopCentre,
+                        Origin = Anchor.TopCentre
+                    };
+                    switch (displayStyle)
+                    {
+                        case PanelDisplayStyle.Grid:
+                            panel.Width = 300;
+                            break;
+                        default:
+                            panel.RelativeSizeAxes = Axes.X;
+                            break;
+                    }
+                    panel.Status.BindTo(u.Status);
+                    return panel;
+                })
             };
 
-            api.Queue(request);
+            LoadComponentAsync(newPanels, p =>
+            {
+                if (panels != null)
+                    ScrollFlow.Remove(panels);
+
+                ScrollFlow.Add(panels = newPanels);
+            });
+        }
+
+        private APIRequest getUsersRequest;
+
+        private readonly Bindable<string> currentQuery = new Bindable<string>();
+
+        private ScheduledDelegate queryChangedDebounce;
+
+        private void updateSearch()
+        {
+            queryChangedDebounce?.Cancel();
+
+            if (!IsLoaded)
+                return;
+
+            Users = null;
+            loading.Hide();
+            getUsersRequest?.Cancel();
+
+            if (api == null || api.State == APIState.Offline)
+                return;
+
+            switch (Header.Tabs.Current.Value)
+            {
+                case SocialTab.OnlinePlayers:
+                    getUsersRequest = new GetUsersRequest(); // TODO filter???
+                    ((GetUsersRequest)getUsersRequest).Success += response => finishRequest(response.Select(r => r.User));
+                    break;
+                case SocialTab.OnlineFriends:
+                    getUsersRequest = new GetFriendsRequest(); // TODO filter???
+                    ((GetFriendsRequest)getUsersRequest).Success += finishRequest;
+                    break;
+            }
+            api.Queue(getUsersRequest);
             loading.Show();
         }
 
+        private void finishRequest(IEnumerable<User> newUsers)

```

Comment:
should probably be called `UpdateUsers` (or just move all this logic to the `Users` setter?)